### PR TITLE
support output logprobs with turbomind backend.

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

Add logprobs output.

~Openai has different logprobs structure of `chat.completions` and `completions` apis, however vllm use same structure of these two api. I think the logprobs structure of `completions` is more user-friendly, so I followed vllm to use this structure with these two apis.~

## Modification

- pytorch / turbomind output with dataclass class
- logprobs logits for turbomind backend

## Use cases (Optional)

```python
from openai import OpenAI
client = OpenAI(base_url='http://0.0.0.0:23333/v1', api_key='sk-l6bdprDovMBW6bRs22B05f5dBa3f417d8bC13e3d131f73Aa')

model_name = client.models.list().data[0].id
completion = client.chat.completions.create(
  model=model_name
  messages=[
    {"role": "user", "content": "讲一个笑话"}
  ],
  logprobs=True,
  top_logprobs=2,
  max_tokens=10,
  # stream=True
)

completion = client.completions.create(
  model=model_name,
  prompt="今天天气真好",
  logprobs=2,
  max_tokens=10,
  # stream=True
)
```

```python

from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig

pipe = pipeline('/nvme/shared/vicuna-7b-v1.5/', backend_config=PytorchEngineConfig())
pipe('hello', gen_config=GenerationConfig(logprobs=10, top_k=40, max_new_tokens=10))
```

### CODE REVIEW

**Comment by lvhan028** [comment]
build failed on windows platform

**Comment by lvhan028** [comment]
May merge latest main to resolve pr_ete_test worflow error

File: lmdeploy/messages.py [hunk 1]
```diff
@@ -212,9 +215,32 @@ class Response:
...
+
+
+@dataclass
+class EngineOutput:
+    """Engine output for turbomind/pytorch engine.
+
+    Args:
+        status (ResponseType): the response type.
+        token_ids (List[int]): the output token ids.
+        num_token (int): the length of output token, for turbomind, num_token
```
Comment (lvhan028) [inline lmdeploy/messages.py:238]: 是可能会多出来一个token么？

File: lmdeploy/serve/openai/protocol.py [hunk 2]
```diff
@@ -61,6 +61,8 @@ class ChatCompletionRequestQos(BaseModel):
     messages: Union[str, List[Dict[str, str]]]
     temperature: Optional[float] = 0.7
     top_p: Optional[float] = 1.0
+    logprobs: Optional[bool] = False
+    top_logprobs: Optional[int] = None
```
Comment (lvhan028) [inline lmdeploy/serve/openai/protocol.py:65]: 有没有上界的限制？

Comment (irexyc) [reply lmdeploy/serve/openai/protocol.py:65]: openai 上限是5。vllm没有限制，turbomind 受限于top_k的kernel，上限是1024 (or 1023)

File: lmdeploy/messages.py (hunk 1 above)
Comment (irexyc) [reply lmdeploy/messages.py:238]: stop word 的时候
https://github.com/InternLM/lmdeploy/blob/e5aaca541f1a73785013b2fc10426082e42c56d8/lmdeploy/turbomind/turbomind.py#L744-L745

**Comment by lvhan028** [comment]
```python
from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig, TurbomindEngineConfig

pipe = pipeline('/workspace/140_models/InternLM/internlm2-chat-7b', backend_config=PytorchEngineConfig())
response = pipe('hello', gen_config=GenerationConfig(logprobs=10, top_k=1, max_new_tokens=10))
print(response)
```
pytorch engine should warn that "logprobs" hasn't been supported yet.

**Comment by lvhan028** [comment]
```python

from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig, TurbomindEngineConfig


pipe = pipeline('/workspace/140_models/InternLM/internlm2-chat-7b')
response = pipe('hello', gen_config=GenerationConfig(logprobs=10, top_k=1, max_new_tokens=10))
print(response)
```
The result is:
```
Response(text='你好！有什么我可以帮助你的吗？', generate_token_len=9, input_token_len=103, session_id=0, finish_reason='stop', token_ids=[77230, 60477, 69259, 74010, 68417, 68364, 61076, 60504], logprobs=[{77230: 0.0}, {60477: 0.0}, {69259: 0.0}, {74010: 0.0}, {68417: 0.0}, {68364: 0.0}, {61076: 0.0}, {60504: 0.0}])
```
1. 为什么 prob的值都是 0.0?
2. generate_token_len 和 token_ids的长度，logprobs 的长度不等。我感觉这会让用户很困惑，我们解释起来也麻烦。能不能在 turbomind.py 中处理好呢？

**Comment by irexyc** [comment]
top_k 为1的话，只有一个候选词，概率是1，log一下就是0了。

logprobs 的长度应该跟token_ids的长度是一样的， 跟generate_token_len长度不一样应该是因为遇到stop word了。我记得这里是为了kv_cache的step吧

top_k 为 2的时候结果是什么？

**Comment by lvhan028** [comment]
从 pipeline 的层面来说，推理时的generate参数，行为，需要和 transformers 一致。
行为一致我们这么规定：
1. 相同的 batch prompt，相同的generate config，如果 batch 中每个prompt通过transformers得到的结果一样，那么lmdeploy也应该一样。不约束和transformers的结果一模一样

**Comment by lvhan028** [comment]
建议增加 ut，对比 lmdeploy 的 logprobs 和 transformers logprobs的结果是否一致。可以把 top_k 设置为1.

File: lmdeploy/serve/openai/protocol.py (hunk 2 above)
Comment (lvhan028) [reply lmdeploy/serve/openai/protocol.py:65]: 那要校验参数的合法性。不要引起crash，hung等严重的问题

File: src/turbomind/triton_backend/llama/LlamaTritonModelInstance.cc [hunk 3]
```diff
@@ -176,6 +177,36 @@ LlamaTritonModelInstance<T>::forward(std::shared_ptr<std::unordered_map<std::str
                                           d_cum_log_probs_}});
     }
 
+    if (input_tensors->count("logprobs")) {
+        size_t max_logprob_length = std::min((int)max_request_output_len, instance_->session_len) + 1;
```
Comment (lvhan028) [inline src/turbomind/triton_backend/llama/LlamaTritonModelInstance.cc:181]: why "+1"

File: src/turbomind/triton_backend/llama/LlamaTritonModelInstance.cc [hunk 4]
```diff
@@ -176,6 +177,36 @@ LlamaTritonModelInstance<T>::forward(std::shared_ptr<std::unordered_map<std::str
...
+        h_logprob_vals_           = (float*)std::realloc(
+            h_logprob_vals_, sizeof(float) * request_batch_size * beam_width * max_logprob_length * 1024);
```
Comment (lvhan028) [inline src/turbomind/triton_backend/llama/LlamaTritonModelInstance.cc:183]: Is the hard value "1024" related to kMaxLogProb?

File: src/turbomind/layers/constant.h [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+
+#pragma once
+
+namespace turbomind {
+
+const int kMaxLogProb = 1024;
```
Comment (lvhan028) [inline src/turbomind/layers/constant.h:7]: kMaxLogProb 是 kernel在用，llamabatch在用，layers不会用。
把这个变量放在 layers 下的头文件中，比较奇怪。
是不是可以把它放在 src/turbomind/utils 下？

File: src/turbomind/models/llama/LlamaV2.cc [hunk 6]
```diff
@@ -372,7 +372,8 @@ void LlamaV2<T>::dynamicDecode(int*            token_ids,
         {"should_stop", {MEMORY_CPU, TYPE_BOOL, {1}, should_stop}},
         {"curand_state", {MEMORY_GPU, TYPE_VOID, {batch_size}, curand_state}}};
 
-    const std::vector<std::string> optional_outputs{"cum_log_probs", "output_log_probs"};
+    const std::vector<std::string> optional_outputs{
+        "cum_log_probs", "output_log_probs", "sampled_indexes", "sampled_logprobs", "sampled_nums"};
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaV2.cc:376]: `cum_log_probs`, `output_log_probs` 是不是没有用到？

**Comment by lzhangzz** [comment]
We need to benchmark the performance impact of requesting logprobs.

**Comment by lvhan028** [comment]
> We need to benchmark the performance impact of requesting logprobs.

internlm2-7b, rps 23.734

**Comment by lvhan028** [comment]
evaluation test pass

**Comment by irexyc** [comment]
with vocab = 92544
image

File: lmdeploy/serve/openai/api_server.py [hunk 7]
```diff
@@ -681,12 +804,22 @@ async def completion_stream_generator() -> AsyncGenerator[str, None]:
                 data = chunk.model_dump_json(exclude_unset=True)
                 yield f'data: {data}\n\n'
 
+            offset = 0
+            all_token_ids = []
+            state = DetokenizeState()
             async for res in generator:
+                logprobs = None
+                if request.logprobs and res.logprobs:
```
[the remark this answers is not in the document]
Comment (irexyc) [reply lmdeploy/serve/openai/api_server.py:812]: 有，pytorch backend的时候

File: src/turbomind/triton_backend/llama/LlamaTritonModelInstance.cc (hunk 3 above)
Comment (irexyc) [reply src/turbomind/triton_backend/llama/LlamaTritonModelInstance.cc:181]: 因为 turbomind的现在的输出总是比request_output_tokens 多一个

**Review by lvhan028 — APPROVED** [review]

**Review by lzhangzz — APPROVED** [review]
